cranelift/isle/isle/src/trie_again.rs

```diff
@@ -188,6 +197,44 @@ pub fn build(termenv: &sema::TermEnv) -> (Vec<(sema::TermId, RuleSet)>, Vec<Erro
     (result, errors)
 }
 
+impl Constraint {
+    /// Return the nested [Binding]s from matching the given [Constraint] against the given [BindingId].
+    pub fn bindings_for(self, source: BindingId) -> impl Iterator<Item = Binding> {
```

Comment:
I think I understand what this method is doing, but it's extremely convoluted and difficult to understand -- the explicit state machine isn't doing it any favors. Since we're inside islec and performance is not extremely critical, could we return a `Vec<Binding>` here? Then one could do a `match self { Constraint::ConstInt { .. } => vec![], Constraint::Some => vec![source], ... }` instead. Likewise this would avoid the need to iterate over the tuple index (above).